### PR TITLE
Allow user to change the channel axis for BatchNorm function and the likes

### PR DESCRIPTION
**Pull request by vboussange**
Proposition for issue #1664

This PR allows the user to customize the channel axis for normalisation functions (`BatchNorm`,` GroupNorm` and `InstanceNorm`).

## Example

```Julia
using Flux
channel_size = 3
channel_axis = 1
BN = BatchNorm(channel_size, dim = channel_axis)
x = randn(channel_size, 10, 10)
BN(x)
```

### CODE REVIEW

File: src/layers/normalise.jl [hunk 1]
```diff
@@ -243,9 +245,11 @@ mutable struct BatchNorm{F,V,N,W}
   track_stats::Bool
   active::Union{Bool, Nothing}
   chs::Int # number of channels
+  dim::Union{Int, Nothing} # channel dimension
```
Comment (DhairyaLGandhi) [inline src/layers/normalise.jl:248]: This should never be `nothing`, but use `N - 1` as default.

Comment (ToucheSir) [reply src/layers/normalise.jl:248]: How do you know `N - 1` a priori though? The question is what is an appropriate sentinel value. Perhaps a negative offset from the end, since by default channels are at dim `end - 1`?

Comment (DhairyaLGandhi) [reply src/layers/normalise.jl:248]: Right so, it will need to be defined at runtime.

**Comment by DhairyaLGandhi** [comment]
How well is cudnn able to handle this?

**Comment by ToucheSir** [comment]
Looping back to answer the cuDNN question: it supports one other configuration for channel dim via `CUDNN_TENSOR_NHWC`. I think adding that to NNlibCUDA at https://github.com/FluxML/NNlibCUDA.jl/blob/96a334633ef3a3707c85fc1754c2c7eb8849db4e/src/cudnn/batchnorm.jl#L27 would be a good first step for getting this going again. Here are a couple example implementation PRs from [MXNet](https://github.com/apache/incubator-mxnet/pull/20615) and [PyTorch](https://github.com/pytorch/pytorch/pull/29361).

**Comment by CarloLucibello** [comment]
Closing as stale: this PR has not seen any new commits in over two years, and Flux has changed substantially since then (Optimisers.jl-based training, the new recurrent layers, Reactant/Enzyme support, etc.), so it can no longer be merged as-is.

This is housekeeping, not a rejection of the idea. If you (or anyone else) would like to pick it up, please feel free to reopen this PR or open a fresh one against the current `master` — the contribution is still appreciated.
